web/src/engine/main/src/keymanEngine.ts

```diff
@@ -152,16 +152,32 @@ export default class KeymanEngine<
         this.osk.startHide(false);
       }
 
-      if(this.osk) {
-        this.osk.setNeedsLayout();
-        this.osk.activeKeyboard = kbd;
-        this.osk.present();
-      }
-
       // Needed to ensure the correct layer is displayed.
       // Needs to be after the OSK has loaded for the keyboard in case the default
       // layer should be something other than "default" for the current context.
-      this.core.resetContext(this.contextManager.activeTarget);
+      const doContextReset = () => {
+        this.contextManager.resetContext();
+      }
+
+      /*
+        This pattern is designed to minimize layout reflow during the keyboard-swap process.
+        The 'default' layer is loaded by default, but some keyboards will start on different
+        layers depending on the current state of the context.
+
+        If possible, we want to only perform layout operations once the correct layer is
+        set to active.
+      */
+      if(this.osk) {
+        this.osk.batchLayoutAfter(() => {
+          this.osk.activeKeyboard = kbd;
+          // Note:  when embedded within the mobile apps, the keyboard will still be visible
+          // at this time.
+          doContextReset();
+          this.osk.present();
+        });
+      } else {
+        doContextReset();
+      }
```

Comment:
I just noticed one relatively easy extra improvement possibly worth considering:

https://github.com/keymanapp/keyman/blob/3cc6305a170aadb4712284b99d3385504063bbbd/web/src/engine/main/src/contextManagerBase.ts#L273-L281

The `resetContext` seen above _could_ be considered redundant.  Granted, the user did interact with some form of UI to change the keyboard setting, so an argument could be made for resetting context in both spots.  That can trigger a layer change depending upon the current context position, and that _would_ have an impact on what the OSK displays should it happen... for the old keyboard, during the transition window to the new keyboard.

The reason I bring it up:

![image](https://github.com/keymanapp/keyman/assets/25213402/8886e7da-37ab-4de1-8dba-362af9bd86a8)

(Note:  this is on _returning_ to a previously-loaded keyboard.)

The earlier call to resetContext, before the keyboard is changed, triggers a few layout operations _and also_ costs about 64.5ms on the test device I was profiling with at the time.